### DISCUSSION TITLE
Prevent Shortcuts from removing the XML response tags?

### DISCUSSION THREAD

**Discussion by charlesboc** [Waboodoo/HTTP-Shortcuts#385] [Q&A] [answered]
I created a simple shortcut to retrieve & update the config of an IP camera using the ISAPI protocol from HikVision.
The API response is in XML. The full XML response body is displayed correctly when the shortcut succeeds.
Problem: when I try reading the `response.body` with a script, the body is a string which contains only the text of the XML response, i.e. all the XML tags have been removed.
Is there a way to get access to the raw XML response body, with all its tags, as it was sent by the server?
I need it to send it back to the server in a subsequent call.
Thank you!

**Comment by Waboodoo** [comment, marked as answer]
Hello,
The `response.body` will actually give you the raw response body as a string. It doesn't remove any tags, so it should already work for what you are trying to achieve. I think the confusion here might stem from the fact that the `alert` / `showDialog` function (which I assume you probably used to inspect the response) interprets its input as HTML (or, a subset thereof). As such, it will not display any of the XML tags that aren't also valid HTML tags that it supports, leading to rather useless and confusing output. The content of `response.body` itself however still contains the tags. You can verify this by using the `logEvent(response.body)` function, which will create an entry in the Event History, which you can find on the app's Settings screen in the Troubleshooting section.
Hope this helps. I'll see that I can improve the way the alert dialog displays text.

**Reply by charlesboc** [reply]
You're completely right, I was using `showDialog` and I missed the fact that it supports HTML. Once HTML encoding is added, I can see the full response body. Thank a lot for the quick reply!

**Comment by adolfousier** [comment]
Hey Alexey, appreciate the deep analysis here. After investigating, the root cause turned out to be different from what the proposal assumes.

TOOLS.md wasn't growing because of structural problems with how it's loaded or indexed. It was growing because the RSI engine (`src/brain/rsi.rs`) was generating suggestions like "Consider adding error handling guidance to TOOLS.md" for every tool with a high failure rate. The autonomous RSI agent followed those suggestions and kept appending failure-derived rules, timestamps, and session references to TOOLS.md.

The failure data already lives in the feedback system (`feedback_record`, `feedback_analyze`). Duplicating it into TOOLS.md created bloat that the dedup scanner then had to clean up.

**What we did:** Changed the RSI suggestion target. Instead of directing the agent to append to TOOLS.md, it now points to `feedback_record` / `feedback_analyze`.

TOOLS.md is for bare minimum tool definitions only: parameter formats, executor types, usage docs. Never mixed with out-of-scope content like failure logs, error notes, or diagnostic data. We had a commit earlier that enforced this separation and moved non-tool-definition content to their own files. The RSI fix reinforces that boundary.

No splitting, no section-level loading, no new filesystem index needed for this specific problem.

That said, the token usage angle you raised is worth digging into further. How and when TOOLS.md gets injected into context is a separate conversation from bloat, and your section-level loading idea could have merit there if the file grows organically as more tools get added. Happy to continue that discussion.
